### DISCUSSION TITLE
Markdown Preview Extra CSS

### DISCUSSION THREAD

**Discussion by Thermal-Shock** [directus/directus#15935] [Archive]
Is it possible to modify the Markdown (input-rich-text-md) preview to include some custom CSS files? 

I realize we are headless but I'd like our content authors to feel a little more comfortable when we initially introduce Directus. In other words, use our theme when showing a Markdown preview. 

I assume I could (a) create my own extension that is a copy of the Directus Markdown interface, but (b) somehow I'd need to tell codemirror to use my extra CSS when running the preview. 

Any rough idea on how I might accomplish (a) and (b)?

**Comment by benhaynes** [comment]
I wonder if you could use the global CSS overrides (in settings) to target that interface, and set your styles natively without any custom extensions. 🤔

**Comment by Thermal-Shock** [comment]
That would be great but my css is too large to fit with the size of the
Custom CSS input field.

If we could upload a file into the file library and then have the Custom
CSS be a list of files in the library, that would be more flexible overall
actually.

Speaking of the file library, it would also be nice to have codemirror
implemented there too so we can edit code files in the file library. Same
as the image edit in the library but for files supported by codemirror. We
could also use it more like  a file system then as well.

All the best.

**Comment by azrikahar** [comment]
Using the Custom CSS in Project Settings would be the most straightforward way to change this at the moment. you can refer to the styles within the `input-rich-text-md` component:

https://github.com/directus/directus/blob/4904c3bba1e360476e316a7d2fc06a5dc10a5409/app/src/interfaces/input-rich-text-md/input-rich-text-md.vue#L584-L587

and omit the `:deep()` usage as that's Vue specific, so it should be something like:

```css
#app .interface-input-rich-text-md .preview-box b,
#app .interface-input-rich-text-md .preview-box strong {
	color: var(--red);
}
```

Here's a quick example:

https://user-images.githubusercontent.com/42867097/195000754-7f6ffe47-9df1-42ce-aa29-35417ecdca35.mp4

> That would be great but my css is too large to fit with the size of the Custom CSS input field.

also note that the Custom CSS interface will expand and become scrollable when there's more content, in case that's the source of your concern. However being able to have a full file editing experience would be nice as well.

> I assume I could (a) create my own extension that is a copy of the Directus Markdown interface, but (b) somehow I'd need to tell codemirror to use my extra CSS when running the preview.
>
> Any rough idea on how I might accomplish (a) and (b)?

The general gist would be you will need to create a [custom interface](https://docs.directus.io/extensions/interfaces.html), copy the `input-rich-text-md` component (and other relative files imported by it) into the new custom interface, and change the all the relevant `.preview-box` classes.

**Reply by Thermal-Shock** [reply]
Thank you for the effort you put into your reply.

The Custom CSS field is just too small to hold all the CSS. I have 4 CSS files that I need to pull in and just pulling in one of them generates the `Value is too long [VALUE_TOO_LONG] Value for field "custom_css" is too long` message from Directus. For reference I'm trying to pull in the CSS for the [Material for Mkdocs](https://squidfunk.github.io/mkdocs-material/) theme and its not very small.

One hacky approach might be for me to upload the CSS files as assets, then somewhere have a piece of JavaScript that simply adds a link to these assets into the DOM head. I'm not sure where I would do this though or if there is another approach, but I really just need to include and link the CSS in the right place? 

If I was to build a custom interface, I'd have a array of CSS assets that I could include.

**Reply by rijkvanzanten** [reply]
> One hacky approach might be for me to upload the CSS files as assets, then somewhere have a piece of JavaScript that simply adds a link to these assets into the DOM head. I'm not sure where I would do this though or if there is another approach, but I really just need to include and link the CSS in the right place?

You might be able to hack that in using the custom CSS 🤔 Try uploading the css file to assets, and then using the following in the custom css:

```css
@import "/assets/<uuid>";
```

**Reply by Thermal-Shock** [reply]
Thanks @rijkvanzanten. This actually works. I really appreciate that.

`@import "/assets/<uuid>";`

This may be obvious due to CSS notation but just in case anyone else sees this, the double-quotes and semi-colon are required.

**Reply by rijkvanzanten** [reply]
> This may be obvious due to CSS notation

It clearly wasn't to me! 😄 I'll update my comment as well to prevent future confusion

**Reply by Thermal-Shock** [reply]
@rijkvanzanten I've run into one wrinkle with this approach. 

If you look at the <head> you can see that the Custom CSS is not included last in the order of stylesheets. 

For me this means that I have to litter my CSS overrides with `!important` for a `h1` for instance since the markdown `.preview-box` already includes `h1` and it gets included by an `index.<hash>.css` after the custom CSS has been included. 

I'd propose that the Custom CSS be included in the head as late as possible by Directus if possible. 

Another quirk you may be aware of is that if I choose Replace File to update my imported CSS file in the file library, it actually still seems to be use a cached version even though I have the cache disabled in the environment. I actually have to delete my CSS file from the library, upload it, and then change the import for the new version to get used.

**Reply by rijkvanzanten** [reply]
> I'd propose that the Custom CSS be included in the head as late as possible by Directus if possible.

It's already as late as it can be. Vue injects the inline styles at the end of the head when the components are rendered. There's currently no way (afaict) to have _another_ <style> locked beneath that.

> Another quirk you may be aware of is that if I choose Replace File to update my imported CSS file in the file library, it actually still seems to be use a cached version even though I have the cache disabled in the environment. I actually have to delete my CSS file from the library, upload it, and then change the import for the new version to get used.

This feels like the browser cache to me. Maybe try clearing the browser cache and refreshing the page?

**Reply by Thermal-Shock** [reply]
Thanks @rijkvanzanten. 

I don't know enough about Vue but I have to believe there is an opportunity to move the Custom CSS. Perhaps the Custom CSS needs to be included as part of a component that is in the page footer then, thereby always making it the last component to be rendered?

I'll play around with the CSS a little more to see if its a cache issue, but I'm thinking I already tried that. I believe I did a replace file, then I opened a new browser tab and used the /assets/<uuid> API to see what it thinks the file contains, and it was still the old one. This should not be subject to browser caching though, correct?

**Reply by rijkvanzanten** [reply]
> This should not be subject to browser caching though, correct?

Depends. If it's a regular tab, you'll have the same browser cache. If it's an incognito window or a different browser, it'll be "Fresh"

**Reply by Thermal-Shock** [reply]
So I did a little more research here and probably not much can be done about this without replacing the file every time. 

If I replace the file, then use the /assets/<uuid> API, it is indeed cached by the browser (chrome), and doing a reload gives me the latest version. This is because the API response headers do not tell the browser not to cache the result. In fact, they say `Cache-Control: public, max-age=2592000` instead. 

However, when I go to the markdown preview in the app, even if I hard reload the page, the app will still use the cached version instead of the new version, as seen below and I cannot force a reload of that one.
![image](https://user-images.githubusercontent.com/66075795/196508844-40abddfa-3a73-45fa-a2c1-2d5134d428f5.png)

My only choice is to remove the file, upload new one, and then modify the @import again. This of course will mean that is cannot be cached and it will use this new one. 

The only solution I see to suggest a feature to Directus to allow setting cache-control or other headers for the assets API under certain circumstances. Similar to the way permissions work. On the app page for my asset in the file library, there would be field to add/replace headers for when the API returns this asset. 

I think this would be good feature in general for any endpoint, whether that be, /assets/ or /items/ on a collection.

**Reply by rijkvanzanten** [reply]
> My only choice is to remove the file, upload new one, and then modify the https://github.com/import again. This of course will mean that is cannot be cached and it will use this new one.

Or you can add a cache buster query parameter to the import in the CSS so you force the browser to download it fresh, for example `@import "/assets/etc?cache-buster=12345"`. The value doesn't matter, as long as you change it to something new once you replace the file

**Reply by Thermal-Shock** [reply]
I still have to update the Custom CSS, but this is certainly more convenient. 

Perhaps also consider adding control over the all endpoint API response headers for completeness though. 

Thanks @rijkvanzanten.
